introduce `SubFPGroup`

ThomasBreuer opened the pull request:
in analogy to `SubPcGroup`, see #3166
(but the situation is not exactly the same because the situation on the GAP side is different)

ThomasBreuer committed:
introduce `SubFPGroup`

fingolfin commented on test/Groups/quotients.jl:
```diff
@@ -92,11 +92,11 @@ end
    @test x == F[1]
    @test y == F[2]
-   @test is_full_fp_group(F)
+   @test Oscar._is_full_fp_group(F)
```
It seems this needs
```suggestion
   @test Oscar._is_full_fp_group(GapObj(F))
```
here and elsewhere -- or alternatively another method which does that for you?

ThomasBreuer replied:
Yes, these were changes in the last moment, after testing other changes ...

ThomasBreuer committed:
add missing `GapObj` indirections

ThomasBreuer committed:
one more missing method extension

ThomasBreuer committed:
fix a call of a not exported function

ThomasBreuer committed:
adjust output of a book example

ThomasBreuer committed:
improve serialization of `SubFPGroup`

- avoid code duplication for `PcGroup`/`SubPcGroup`, `FPGroup`, `ScubFPGroup`
- change serialization format for `(Sub)FPGroup`, `(Sub)PcGroup`:
  we need more references for relations between (sub)groups
- add tests for this situation
- document `SubFPGroup` as serializable
- call `GapObj` instead of accessing `.X`
- add some `GAPWrap` variables

ThomasBreuer committed:
add `full_group`

for both `SubPcGroup` and `SubFPGroup`

(and fix a typo in the documentation)

lgoettgens commented on src/Groups/GAPGroups.jl:
````diff
@@ -1741,31 +1741,34 @@ false
 ```
 """
-is_full_fp_group(G::FPGroup) = GAPWrap.IsFpGroup(GapObj(G))
+function full_group(G::T) where T <: Union{SubFPGroup, SubPcGroup}
````
```suggestion
function full_group(G::Union{SubFPGroup, SubPcGroup})
```
If you don't use `T` in the body, it is more efficient (in terms of compile time) to not have the type parameter at all

lgoettgens commented on src/Groups/GAPGroups.jl:
```diff
@@ -1741,31 +1741,34 @@ false
+end
+
+# for convenience
+function full_group(G::T) where T <: Union{FPGroup, PcGroup}
```
```suggestion
function full_group(G::Union{FPGroup, PcGroup})
```

lgoettgens commented on src/Groups/GAPGroups.jl:
```diff
@@ -2237,7 +2235,7 @@ function describe(G::GAPGroup)
 end
 
-function describe(G::FPGroup)
+function describe(G::T) where T <: Union{FPGroup, SubFPGroup}
```
```suggestion
function describe(G::Union{FPGroup, SubFPGroup})
```

ThomasBreuer committed:
address comments

ThomasBreuer commented:
The failure in `test (1.10, book, ubuntu-latest` is due to a GC error.
```
...
GC: pause 73.72ms. collected 499.818481MB. incr 
    explSL25.jlcon
GC error (probable corruption)
Allocations: 575920478 (Pool: 575437139; Big: 483339); GC: 110
<?#0x7f7ef198c000::(nil)>

thread 0 ptr queue:
~~~~~~~~~~ ptr queue top ~~~~~~~~~~
GAP.SmallBag()
==========
...
```

ThomasBreuer commented:
Now I get a failure in `Run CI without experimental / doctest` but I cannot open the section "Run doctests" which is marked as failed.

lgoettgens commented:
> Now I get a failure in `Run CI without experimental / doctest` but I cannot open the section "Run doctests" which is marked as failed.

possibly some github issue. I restarted the job, maybe we can see the logs now

benlorenz commented:
Sometimes it helps opening the [raw log output](https://github.com/oscar-system/Oscar.jl/commit/60aba0a86366a06f0750f93e40d12425ddff09f5/checks/25450591858/logs) via the button on the top right, in this case it shows a GAP error during a `gmodule` call:
````
Error: doctest failure in ~/work/Oscar.jl/Oscar.jl/experimental/GModule/GaloisCohomology.jl:1501-1523

```jldoctest

julia> G = SL(2,5)
SL(2,5)

julia> T = character_table(G);

julia> R = gmodule(T[9])
G-module for G acting on vector space of dimension 6 over abelian closure of Q

julia> S = gmodule(CyclotomicField, R)
G-module for G acting on vector space of dimension 6 over cyclotomic field of order 5

julia> B, mB = relative_brauer_group(base_ring(S), character_field(S));

julia> B
Relative Brauer group for cyclotomic field of order 5 over number field of degree 1 over QQ

julia> b = B(S)
Element of relative Brauer group of number field of degree 1 over QQ
  <2, 2> -> 1//2 + Z
  <5, 5> -> 0 + Z
  Complex embedding of number field -> 1//2 + Z
```

Subexpression:

R = gmodule(T[9])

Evaluated output:

ERROR: Error thrown by GAP: Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `*' on 2 arguments at /home/runner/.julia/artifacts/b5c2f0f824457e5c391fb24916f94d5d91c62c4f/share/gap/lib/methsel2.g:249 called from
PROD_SCL_LIST_DEFAULT( nonlist, list ) at /home/runner/.julia/artifacts/b5c2f0f824457e5c391fb24916f94d5d91c62c4f/share/gap/lib/list.gi:3433 called from
PROD_SCL_LIST_DEFAULT( nonlist, list ) at /home/runner/.julia/artifacts/b5c2f0f824457e5c391fb24916f94d5d91c62c4f/share/gap/lib/list.gi:3433 called from
u * GeneratorsOfGroup( S[i] ) at /home/runner/.julia/scratchspaces/c863536a-3901-11e9-33e7-d5cd0df7b904/gap_1387324105167574757_1.10/pkg/repsn/gap/repsn.g:1389 called from
func( C[i] ) at /home/runner/.julia/artifacts/b5c2f0f824457e5c391fb24916f94d5d91c62c4f/share/gap/lib/coll.gi:663 called from
List( Co, function ( u )
return u * GeneratorsOfGroup( S[i] );
end ) at /home/runner/.julia/scratchspaces/c863536a-3901-11e9-33e7-d5cd0df7b904/gap_1387324105167574757_1.10/pkg/repsn/gap/repsn.g:1389 called from
...  at *defin*:0

Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] ThrowObserver(depth::Int32)
   @ GAP ~/.julia/packages/GAP/fMeZB/src/GAP.jl:89
 [3] _call_gap_func
   @ ~/.julia/packages/GAP/fMeZB/src/ccalls.jl:330 [inlined]
 [4] (::GapObj)(a1::GapObj)
   @ GAP ~/.julia/packages/GAP/fMeZB/src/ccalls.jl:307
 [5] gmodule(chi::Oscar.GAPGroupClassFunction)
   @ Oscar.GModuleFromGap ~/work/Oscar.jl/Oscar.jl/experimental/GModule/GModule.jl:1531
 [6] top-level scope
   @ none:1
````

ThomasBreuer commented:
Thanks @benlorenz .
This error looks new ...
